File: tensorflow/lite/delegates/gpu/common/shape_test.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow/lite/delegates/gpu/common/shape.h"

#include <initializer_list>
#include <vector>

#include <gmock/gmock.h>
#include <gtest/gtest.h>

namespace tflite {
namespace gpu {
namespace {

TEST(OIHW, Smoke) {
  OIHW OIHW;

  // Test 4 different versions of setters.
  OIHW.i = 1;
  ASSERT_TRUE(OIHW.set<Axis::OUTPUT_CHANNELS>(2));
  ASSERT_TRUE(OIHW.set(Axis::HEIGHT, 3));
  ASSERT_TRUE(OIHW.set(3, 4));

  // Make sure invalid setters return false.
  ASSERT_FALSE(OIHW.set(5, 10));
  ASSERT_FALSE(OIHW.set(Axis::CHANNELS, 10));
  ASSERT_FALSE(OIHW.set<Axis::CHANNELS>(10));

  // Test 4 different versions of getters
  EXPECT_EQ(1, OIHW.get(Axis::INPUT_CHANNELS));
  EXPECT_EQ(2, OIHW.o);
  EXPECT_EQ(3, OIHW.get(2));
  EXPECT_EQ(4, OIHW.get<Axis::WIDTH>());

  // Make sure getters that fall outside of a range return invalid axis.
  EXPECT_EQ(-1, OIHW.get(5));
  EXPECT_EQ(-1, OIHW.get(Axis::CHANNELS));
  EXPECT_EQ(-1, OIHW.get<Axis::CHANNELS>());

  // Check axis indices are all correct.
  ASSERT_EQ(4, OIHW.size());
  std::vector<Axis> expected = {Axis::OUTPUT_CHANNELS, Axis::INPUT_CHANNELS,
                                Axis::HEIGHT, Axis::WIDTH};
  for (int i = 0; i < OIHW.size(); ++i) {
    Axis axis = OIHW.axis(i);
    ASSERT_EQ(expected[i], axis);
    ASSERT_EQ(i, OIHW.index(axis));
  }

  // Check equivalent conversions.
  OHWI ohwi;
  ASSERT_TRUE(ohwi.CopyAllDefinedAxis(OIHW));
  EXPECT_EQ(ohwi.o, OIHW.o);
  EXPECT_EQ(ohwi.i, OIHW.i);
  EXPECT_EQ(ohwi.h, OIHW.h);
  EXPECT_EQ(ohwi.w, OIHW.w);

  ohwi = OHWI(10, 20, 30, 40);
  ASSERT_TRUE(OIHW.CopyAllGivenAxis(ohwi));
  EXPECT_EQ(ohwi.o, OIHW.o);
  EXPECT_EQ(ohwi.i, OIHW.i);
  EXPECT_EQ(ohwi.h, OIHW.h);
  EXPECT_EQ(ohwi.w, OIHW.w);

  EXPECT_TRUE(ohwi.has(Axis::WIDTH));
  EXPECT_FALSE(ohwi.has(Axis::DEPTH));
}

TEST(Layout, Smoke) {
  EXPECT_EQ(4, Size<Layout::OIHW>());
  EXPECT_EQ(4, Size(Layout::OIHW));
  std::vector<Axis> expected = {Axis::OUTPUT_CHANNELS, Axis::INPUT_CHANNELS,
                                Axis::HEIGHT, Axis::WIDTH};
  for (int i = 0; i < Size<Layout::OIHW>(); ++i) {
    Axis axis = GetAxis<Layout::OIHW>(i);
    ASSERT_EQ(expected[i], axis);
    ASSERT_EQ(axis, GetAxis(Layout::OIHW, i));
    ASSERT_EQ(i, GetAxisIndex<Layout::OIHW>(axis));
    ASSERT_EQ(i, GetAxisIndex(Layout::OIHW, axis));
  }
  EXPECT_EQ(Axis::UNKNOWN, GetAxis(Layout::OIHW, 5));
  EXPECT_EQ(-1, GetAxisIndex<Layout::OIHW>(Axis::CHANNELS));
  EXPECT_EQ(-1, GetAxisIndex<Layout::OIHW>(Axis::CHANNELS));
  EXPECT_TRUE(HasAxis<Layout::OHWDI>(Axis::DEPTH));
  EXPECT_FALSE(HasAxis<Layout::OHWDI>(Axis::CHANNELS));
}

TEST(Shape, Smoke) {
  Shape s(Layout::OIHW, {1, 2, 3, 4});
  EXPECT_TRUE(s.set(Axis::HEIGHT, 10));
  EXPECT_TRUE(s.set<Axis::WIDTH>(20));
  EXPECT_FALSE(s.set(Axis::BATCH, 10));
  EXPECT_FALSE(s.set<Axis::BATCH>(20));

  ASSERT_EQ(10, s.get<Axis::HEIGHT>());
  ASSERT_EQ(20, s.get(Axis::WIDTH));
  EXPECT_EQ(20, s.dimensions[3]);

  EXPECT_TRUE(s.has(Axis::HEIGHT));
  EXPECT_FALSE(s.has(Axis::DEPTH));

  OIHW oihw(1, 2, 10, 20);
  Shape s2 = oihw.ToShape();
  EXPECT_EQ(s2.layout, oihw.layout);
  EXPECT_EQ(s.layout, s2.layout);
  EXPECT_EQ(s.dimensions, s2.dimensions);

  // Convert layout into compatible shape.
  OHWI ohwi;
  ASSERT_TRUE(ohwi.Adopt(s2));
  EXPECT_EQ(1, ohwi.o);
  EXPECT_EQ(2, ohwi.i);
  EXPECT_EQ(10, ohwi.h);
  EXPECT_EQ(20, ohwi.w);
}

}  // namespace
}  // namespace gpu
}  // namespace tflite

File: tensorflow/go/shape_test.go
/*
Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package tensorflow

import (
	"fmt"
	"reflect"
	"testing"
)

func TestShape(t *testing.T) {
	tests := []struct {
		shape Shape
		slice []int64
		full  bool
		str   string
	}{
		{
			shape: ScalarShape(),
			slice: make([]int64, 0),
			full:  true,
			str:   "[]",
		},
		{
			shape: MakeShape(-1, 2, -1, 4),
			slice: []int64{-1, 2, -1, 4},
			full:  false,
			str:   "[?, 2, ?, 4]",
		},
		{
			shape: MakeShape(2, 3),
			slice: []int64{2, 3},
			full:  true,
			str:   "[2, 3]",
		},
	}
	for _, test := range tests {
		t.Run(fmt.Sprintf("%#v", test.shape), func(t *testing.T) {
			if got, want := test.shape.NumDimensions(), len(test.slice); got != want {
				t.Errorf("Got %v, want %v", got, want)
			}
			if gotSlice, err := test.shape.ToSlice(); err != nil || !reflect.DeepEqual(gotSlice, test.slice) {
				t.Errorf("Got (%#v, %v), want (%#v, nil)", gotSlice, err, test.slice)
			}
			if got, want := test.shape.IsFullySpecified(), test.full; got != want {
				t.Errorf("Got %v, want %v", got, want)
			}
			if got, want := test.shape.String(), test.str; got != want {
				t.Errorf("Got %v, want %v", got, want)
			}
		})
	}

}

func TestZeroShape(t *testing.T) {
	var s Shape
	if s.NumDimensions() != -1 {
		t.Error(s.NumDimensions())
	}
	if _, err := s.ToSlice(); err == nil {
		t.Error("ToSlice() on a Shape of unknown number of dimensions should fail")
	}
	if s.IsFullySpecified() {
		t.Error("Shape of unknown number of dimensions should not be fully specified")
	}
	if got, want := s.String(), "?"; got != want {
		t.Errorf("Got %q, want %q", got, want)
	}

}
